tracker.go

```diff
@@ -32,98 +34,160 @@ func (c *Config) SetDefault() {
+		if doCapture {
+			t.lastCapture = now
 		}
-		if now.Sub(t.lastReport) < t.config.reportInterval {
-			t.writeReport(now)
+		t.lastCaptureMutex.Unlock()
+
+		if doCapture {
+			t.captureTrace()
 		}
 	}
 }
 
 func (t *StackTracker) captureTrace() {
-	var pc []uintptr = make ([]uintptr, 64)
+	var pc []uintptr = make([]uintptr, 64)
 	l := runtime.Callers(2, pc)
 	pc = pc[:l]
 	key := fmt.Sprintf("%X", pc)
 	// trim brackets from string of form "[0123 4567 89ab cdef]"
-	key = key[1:len(key)-1]
-	t.traces[key]+= 1
+	key = key[1 : len(key)-1]
+
+	t.tracesMutex.RLock()
+	if t.traceChannel != nil {
+		t.traceChannel <- key
+	}
+	t.tracesMutex.RUnlock()
+
+}
+
+func (t *StackTracker) intervalTester() {
+	intervalStartTime := time.Now()
+	intervalStartCount := int32(0)
+	var finalizeReporter chan interface{}
+	for {
+		time.Sleep(t.config.interval)
+		now := time.Now()
+		deltaT := now.Sub(intervalStartTime)
+		deltaC := atomic.LoadInt32(&t.count) - intervalStartCount
+		var rate float64 = float64(deltaC) * float64(time.Second) / float64(deltaT)
+		if !t.isTracking() {
+			if rate > t.config.thresholdStartRate {
+				finalizeReporter = t.startTracking()
+			}
+		} else {
+			if rate < t.config.thresholdStopRate {
+				t.stopTracking(finalizeReporter)
+				finalizeReporter = nil
+			}
+		}
+
+		intervalStartTime = now
+		intervalStartCount = t.count
+	}
 }
 
 func (t *StackTracker) isTracking() bool {
-	return t.traces != nil
+	t.tracesMutex.RLock()
+	defer t.tracesMutex.RUnlock()
+	return t.traceChannel != nil
 }
-func (t *StackTracker) startTracking(now time.Time) {
-	t.traces = make(map[string]uint32)
-	t.lastReport = now
+
+func (t *StackTracker) startTracking() chan interface{} {
+	traceChannel := make(chan string, 128)
+	finalizeReporter := make(chan interface{})
+	go reportWriter(t.config, traceChannel, finalizeReporter)
+	t.tracesMutex.Lock()
+	t.traceChannel = traceChannel
+	t.tracesMutex.Unlock()
+	return finalizeReporter
 }
-func (t *StackTracker) stopTracking(now time.Time) {
-	t.writeReport(now)
-	t.traces = nil
+
+func (t *StackTracker) stopTracking(finalizeReporter chan interface{}) {
+	t.tracesMutex.Lock()
+	t.traceChannel = nil
+	t.tracesMutex.Unlock()
+	close(finalizeReporter)
 }
 
+type traceMap map[string]uint32
+
+func reportWriter(config *Config, traceChan chan string, final chan interface{}) {
+	traces := make(traceMap)
+	timer := time.After(config.reportInterval)
+	for {
+		select {
+		case key := <-traceChan:
+			traces[key] += 1
+		case <-timer:
+			tmp := traces
+			traces = make(traceMap)
+			timer = time.After(config.reportInterval)
+			go writeReport(tmp, config.writer)
+		case <-final:
+		ReadAllTraces:
+			for {
+				select {
+				case key := <-traceChan:
+					traces[key] += 1
+				default:
+					break ReadAllTraces
+				}
+			}
 
-func (t *StackTracker) writeReport(now time.Time) {
-	t.lastReport = now
-	var buffer bytes.Buffer
+			writeReport(traces, config.writer)
+			return
+		}
+	}
+}
```

Comment:
Sure.  The routine takes a map [stacktrace]count and copies it into a list of (count, stacktrace) tuples.  Then it sorts the list by count and outputs it in order.

The long term intent is to allow some configuration - e.g., only print the top three, or any that are above 10%, or something.